File: rerun-reports/02-python/20260829-032449/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/azure-skill-mcp/generated-code/blob_event_notifier/event_publisher.py
"""Sync and async publishers for custom downstream Event Grid events."""

from __future__ import annotations

import logging
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

from azure.core.exceptions import AzureError
from azure.eventgrid import EventGridEvent

from .config import (
    create_async_event_grid_publisher_client,
    create_event_grid_publisher_client,
)

LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True)
class CustomEvent:
    event_type: str
    data: Mapping[str, Any]
    subject: str
    data_version: str = "1.0"


def _normalize_subject(subject: str) -> str:
    normalized = "/" + subject.strip("/")
    if normalized == "/" or "//" in normalized:
        raise ValueError("Event subject must contain a valid hierarchy")
    return normalized


def _to_event_grid_events(events: Sequence[CustomEvent]) -> list[EventGridEvent]:
    if not events:
        raise ValueError("At least one custom event is required")
    return [
        EventGridEvent(
            subject=_normalize_subject(event.subject),
            event_type=event.event_type,
            data=dict(event.data),
            data_version=event.data_version,
        )
        for event in events
    ]


def publish_events(
    endpoint: str,
    events: Sequence[CustomEvent],
    *,
    client: Any | None = None,
) -> bool:
    publisher = client or create_event_grid_publisher_client(endpoint)
    owns_client = client is None
    try:
        publisher.send(_to_event_grid_events(events))
        LOGGER.info("Published %d downstream event(s)", len(events))
        return True
    except AzureError:
        LOGGER.exception("Event Grid publishing failed")
        return False
    finally:
        if owns_client:
            publisher.close()


async def publish_events_async(
    endpoint: str,
    events: Sequence[CustomEvent],
    *,
    client: Any | None = None,
) -> bool:
    publisher = client or create_async_event_grid_publisher_client(endpoint)
    owns_client = client is None
    try:
        await publisher.send(_to_event_grid_events(events))
        LOGGER.info("Published %d downstream event(s)", len(events))
        return True
    except AzureError:
        LOGGER.exception("Event Grid publishing failed")
        return False
    finally:
        if owns_client:
            await publisher.close()

File: rerun-reports/03-java/20260829-075759/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp/generated-code/src/main/java/com/example/blobevents/EventPublisher.java
package com.example.blobevents;

import com.azure.core.credential.TokenCredential;
import com.azure.core.util.BinaryData;
import com.azure.messaging.eventgrid.EventGridEvent;
import com.azure.messaging.eventgrid.EventGridPublisherClient;
import com.azure.messaging.eventgrid.EventGridPublisherClientBuilder;

import java.util.List;
import java.util.Objects;

public final class EventPublisher {
    private final SyncEventSender sender;

    public EventPublisher(String topicEndpoint, TokenCredential credential) {
        EventGridPublisherClient<EventGridEvent> client = new EventGridPublisherClientBuilder()
            .endpoint(topicEndpoint)
            .credential(credential)
            .buildEventGridEventPublisherClient();
        this.sender = client::sendEvents;
    }

    EventPublisher(SyncEventSender sender) {
        this.sender = Objects.requireNonNull(sender, "sender");
    }

    public void publish(List<CustomEvent> events) {
        sender.send(toEventGridEvents(events));
    }

    static List<EventGridEvent> toEventGridEvents(List<CustomEvent> events) {
        Objects.requireNonNull(events, "events");
        if (events.isEmpty()) {
            throw new IllegalArgumentException("At least one custom event is required");
        }
        return events.stream()
            .map(event -> new EventGridEvent(
                event.subject(),
                event.eventType(),
                BinaryData.fromObject(event.data()),
                event.dataVersion()))
            .toList();
    }

    @FunctionalInterface
    interface SyncEventSender {
        void send(Iterable<EventGridEvent> events);
    }
}
